### PR TITLE
[1/N] Avoid copy in std::get 

### PR DESCRIPTION
**Pull request by cyyever**
Fixes #ISSUE_NUMBER

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/functorch/BatchRulesBinaryOps.cpp [hunk 1]
```diff
@@ -14,17 +14,15 @@
...
 
-  auto tensor_other = _binary_pointwise_helper(
+  auto [tensor_, other_]= _binary_pointwise_helper(
       tensor, tensor_batch_dim, other, other_batch_dim);
-  auto tensor_ = std::get<0>(tensor_other);
-  auto other_ = std::get<1>(tensor_other);
 
-  auto result = Func(tensor_, other_, std::forward<ExtraArgs>(extra_args)...);
+  auto result = Func(tensor_, std::move(other_), std::forward<ExtraArgs>(extra_args)...);
   return std::make_tuple(result, 0);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesBinaryOps.cpp:26]: Nit: should the result be moved too?

File: aten/src/ATen/functorch/BatchRulesHelper.h [hunk 2]
```diff
@@ -265,13 +264,9 @@ inline void boxed_existing_bdim_all_batch_rule(
 
   // for each tensor, ensure it has a bdim and reshape it.
   for (const auto tensor_idx : c10::irange(0, tensor_inputs.size())) {
-    const auto& value = std::get<0>(tensor_inputs[tensor_idx]);
-    auto bdim = std::get<1>(tensor_inputs[tensor_idx]);
+    const auto& [value,bdim] = tensor_inputs[tensor_idx];
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesHelper.h:267]: Huh, clang format doesn't complain about the lack of space here between the two parts of the structured assignment.

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp [hunk 3]
```diff
@@ -236,20 +236,16 @@ struct LinalgCheckMatrixBinaryRuleHelper<op_name, F, Func, typelist<A, B, T...>>
...
   static twoOutputs apply_two(
       const Tensor& first, std::optional<int64_t> first_bdim,
       const Tensor& second, std::optional<int64_t> second_bdim,
       T... extra_args) {
-    const auto tensor_other = check_inputs_and_reshape_inputs(first, first_bdim, second, second_bdim);
-    const auto tensor_ = std::get<0>(tensor_other);
-    const auto other_ = std::get<1>(tensor_other);
-    const auto res = Func(tensor_, other_, std::forward<T>(extra_args)...);
+    auto [tensor_, other_]= check_inputs_and_reshape_inputs(first, first_bdim, second, second_bdim);
+    const auto res = Func(std::move(tensor_), std::move(other_), std::forward<T>(extra_args)...);
```
Comment (Skylion007) [inline aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:248]: Don't both to move the res here on the next line?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -625,17 +625,6 @@ template Tensor div_tensor_self_backward(
     ScalarType,
     const std::optional<c10::string_view>&);
 
-template <typename T>
```
Comment (Skylion007) [inline torch/csrc/autograd/FunctionsManual.cpp:628]: Unused I'm guessing?

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/functorch/BatchRulesHelper.h (hunk 2 above)
Comment (cyyever) [reply aten/src/ATen/functorch/BatchRulesHelper.h:267]: The file is excluded

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 4 above)
Comment (cyyever) [reply torch/csrc/autograd/FunctionsManual.cpp:628]: Yes

File: aten/src/ATen/functorch/BatchRulesBinaryOps.cpp (hunk 1 above)
Comment (cyyever) [reply aten/src/ATen/functorch/BatchRulesBinaryOps.cpp]: Fixed

File: aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp (hunk 3 above)
Comment (cyyever) [reply aten/src/ATen/functorch/BatchRulesLinearAlgebra.cpp:248]: Fixed

**Pull request closed without merging**